bindings/python/pymongoarrow/bson.pyi

```diff
@@ -13,56 +13,69 @@
 # limitations under the License.
 
 
+cdef const bson_t* bson_reader_read_safe(bson_reader_t* stream_reader, cbool* reached_eof) except? NULL:
+    cdef const bson_t* doc = bson_reader_read(stream_reader, reached_eof)
+    if doc == NULL and dereference(reached_eof) == False:
+        raise RuntimeError("Could not read BSON stream")
+    return doc
+
+
 def process_bson_stream(bson_stream, context):
     cdef const uint8_t* docstream = <const uint8_t *>bson_stream
     cdef size_t length = <size_t>PyBytes_Size(bson_stream)
     cdef bson_reader_t* stream_reader = bson_reader_new_from_data(docstream, length)
     cdef cbool reached_eof = False
-    cdef const bson_t * doc = bson_reader_read(stream_reader, & reached_eof)
+    cdef const bson_t * doc = bson_reader_read_safe(stream_reader, &reached_eof)
     cdef bson_iter_t doc_iter
     cdef const char* key
     cdef bson_type_t value_t
+    cdef uint64_t count = 0
 
     builder_map = context.builder_map
     type_map = context.type_map
 
-    while doc != NULL:
-        if not bson_iter_init(&doc_iter, doc):
-            raise RuntimeError
-        while bson_iter_next(&doc_iter):
-            key = bson_iter_key(&doc_iter)
-            if key in builder_map:
-                builder = builder_map[key]
-                ftype = type_map[key]
-                value_t = bson_iter_type(&doc_iter)
-                if ftype == _BsonArrowTypes.int32:
-                    if value_t == BSON_TYPE_INT32:
-                        builder.append(bson_iter_int32(&doc_iter))
-                    else:
-                        builder.append_null()
-                elif ftype == _BsonArrowTypes.int64:
-                    if (value_t == BSON_TYPE_INT64 or
-                            value_t == BSON_TYPE_BOOL or
-                            value_t == BSON_TYPE_DOUBLE or
-                            value_t == BSON_TYPE_INT32):
-                        builder.append(bson_iter_as_int64(&doc_iter))
+    try:
+        while doc != NULL:
```

Comment:
Nit: I suggest putting the first call to bson_reader_read_safe inside the try/except. We could even refactor to have a single call instead of 2:
```
    try:
        while True:
            doc = bson_reader_read_safe(stream_reader, &reached_eof)
            if doc == NULL:
                break
            if not bson_iter_init(&doc_iter, doc):....
```

Reply:
Nice! Done.